File: ios/Fly/Core/Transfer/TransferManager.swift
import Foundation
import Combine

class TransferManager: ObservableObject {
    @Published var transfers: [TransferItem] = []

    private var pendingChunks: [String: [Data]] = [:]
    private var pendingFileInfos: [String: (name: String, mime: String, size: Int64, totalChunks: Int)] = [:]
    private var cancellables = Set<AnyCancellable>()

    var chunkThresholdBytes: Int64 = 64 * 1024 * 1024
    var chunkSizeBytes: Int = 4 * 1024 * 1024

    func addConnection(_ conn: PeerConnection, deviceName: String) {
        conn.incomingMessages
            .receive(on: DispatchQueue.main)
            .sink { [weak self] (msg, body) in
                self?.handle(msg, body: body, deviceName: deviceName, conn: conn)
            }
            .store(in: &cancellables)
    }

    private func handle(_ msg: FlyMessage, body: Data, deviceName: String, conn: PeerConnection) {
        switch msg {
        case .hello:
            break

        case .text(let id, let content):
            prepend(.text(id: id, deviceName: deviceName, isIncoming: true, content: content))

        case .fileOffer(let id, let name, let mime, let size):
            let url = saveFile(name: name, data: body)
            prepend(.file(id: id, deviceName: deviceName, isIncoming: true, name: name, mime: mime, size: size, localURL: url, progress: 1.0, totalChunks: 1, receivedChunks: 1))
            conn.send(.ack(id: id))

        case .fileStart(let id, let name, let mime, let size, let totalChunks, _):
            pendingFileInfos[id] = (name, mime, size, totalChunks)
            pendingChunks[id] = []
            prepend(.file(id: id, deviceName: deviceName, isIncoming: true, name: name, mime: mime, size: size, localURL: nil, progress: 0.0, totalChunks: totalChunks, receivedChunks: 0))

        case .chunk(let id, _):
            pendingChunks[id]?.append(body)
            guard let info = pendingFileInfos[id] else { break }
            let received = pendingChunks[id]?.count ?? 0
            updateFile(id: id, progress: Double(received) / Double(info.totalChunks), received: received)

        case .fileDone(let id):
            guard let chunks = pendingChunks.removeValue(forKey: id),
                  let info = pendingFileInfos.removeValue(forKey: id) else { break }
            let fullData = chunks.reduce(Data()) { $0 + $1 }
            let url = saveFile(name: info.name, data: fullData)
            updateFile(id: id, progress: 1.0, received: info.totalChunks, localURL: url)
            conn.send(.ack(id: id))

        case .logcat(let id, let line):
            if let idx = transfers.firstIndex(where: { $0.id == id }),
               case .logcatStream(_, let dn, let inc, let lines) = transfers[idx] {
                transfers[idx] = .logcatStream(id: id, deviceName: dn, isIncoming: inc, lines: lines + [line])
            } else {
                prepend(.logcatStream(id: id, deviceName: deviceName, isIncoming: true, lines: [line]))
            }

        default:
            break
        }
    }

    func sendText(_ text: String, via conn: PeerConnection) {
        let id = UUID().uuidString
        conn.send(.text(id: id, content: text))
        prepend(.text(id: id, deviceName: "Me", isIncoming: false, content: text))
    }

    func sendFile(at url: URL, via conn: PeerConnection) {
        let id = UUID().uuidString
        let name = url.lastPathComponent
        guard let data = try? Data(contentsOf: url) else { return }
        let size = Int64(data.count)
        let mime = "application/octet-stream"

        if size <= chunkThresholdBytes {
            conn.send(.fileOffer(id: id, name: name, mime: mime, size: size), body: data)
            prepend(.file(id: id, deviceName: "Me", isIncoming: false, name: name, mime: mime, size: size, localURL: url, progress: 1.0, totalChunks: 1, receivedChunks: 1))
        } else {
            let chunkSize = chunkSizeBytes
            let totalChunks = Int(ceil(Double(size) / Double(chunkSize)))
            conn.send(.fileStart(id: id, name: name, mime: mime, size: size, totalChunks: totalChunks, chunkSize: chunkSize))
            prepend(.file(id: id, deviceName: "Me", isIncoming: false, name: name, mime: mime, size: size, localURL: url, progress: 0.0, totalChunks: totalChunks, receivedChunks: 0))
            DispatchQueue.global(qos: .utility).async { [weak self] in
                var offset = 0
                var index = 0
                while offset < data.count {
                    let end = min(offset + chunkSize, data.count)
                    let chunk = data[offset..<end]
                    conn.send(.chunk(id: id, index: index), body: Data(chunk))
                    let received = index + 1
                    DispatchQueue.main.async {
                        self?.updateFile(id: id, progress: Double(received) / Double(totalChunks), received: received)
                    }
                    offset = end
                    index += 1
                }
                conn.send(.fileDone(id: id))
                DispatchQueue.main.async { self?.updateFile(id: id, progress: 1.0, received: totalChunks) }
            }
        }
    }

    private func saveFile(name: String, data: Data) -> URL {
        let dir = FileManager.default.temporaryDirectory.appendingPathComponent("fly_received", isDirectory: true)
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        let url = dir.appendingPathComponent(name)
        try? data.write(to: url)
        return url
    }

    private func prepend(_ item: TransferItem) {
        transfers.insert(item, at: 0)
    }

    private func updateFile(id: String, progress: Double, received: Int, localURL: URL? = nil) {
        if let idx = transfers.firstIndex(where: { $0.id == id }),
           case .file(_, let dn, let inc, let name, let mime, let size, let existingURL, _, let total, _) = transfers[idx] {
            transfers[idx] = .file(id: id, deviceName: dn, isIncoming: inc, name: name, mime: mime, size: size, localURL: localURL ?? existingURL, progress: progress, totalChunks: total, receivedChunks: received)
        }
    }
}

File: android/app/src/main/java/com/fly/app/core/transfer/TransferManager.kt
package com.fly.app.core.transfer

import android.content.ContentResolver
import android.content.Context
import android.net.Uri
import android.provider.OpenableColumns
import com.fly.app.core.network.PeerConnection
import com.fly.app.core.protocol.FlyMessage
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.io.File
import java.util.UUID
import kotlin.math.ceil

class TransferManager(
    private val context: Context,
    private val scope: CoroutineScope,
    private val chunkThresholdBytes: () -> Long,
    private val chunkSizeBytes: () -> Int,
    private val deviceName: () -> String = { "My Phone" }
) {
    private val _transfers = MutableStateFlow<List<TransferItem>>(emptyList())
    val transfers: StateFlow<List<TransferItem>> = _transfers.asStateFlow()

    private val pendingChunks = mutableMapOf<String, MutableList<ByteArray>>()
    private val pendingFileInfos = mutableMapOf<String, FlyMessage.FileStart>()

    fun addConnection(conn: PeerConnection, deviceName: String) {
        var actualDeviceName = deviceName
        scope.launch {
            conn.incoming.collect { (msg, body) ->
                if (msg is FlyMessage.Hello && actualDeviceName == "incoming device") {
                    actualDeviceName = msg.deviceName
                }
                handleIncoming(msg, body, actualDeviceName, conn)
            }
        }
    }

    private suspend fun handleIncoming(msg: FlyMessage, body: ByteArray, deviceName: String, conn: PeerConnection) {
        when (msg) {
            is FlyMessage.Hello -> {}

            is FlyMessage.Text -> prepend(
                TransferItem.Text(msg.id, deviceName, true, msg.content)
            )

            is FlyMessage.FileOffer -> {
                val file = saveFile(msg.name, body)
                prepend(TransferItem.File(msg.id, deviceName, true, msg.name, msg.mime, msg.size, file.absolutePath, 1f))
                conn.send(FlyMessage.Ack(msg.id))
            }

            is FlyMessage.FileStart -> {
                pendingFileInfos[msg.id] = msg
                pendingChunks[msg.id] = mutableListOf()
                prepend(TransferItem.File(msg.id, deviceName, true, msg.name, msg.mime, msg.size, null, 0f, msg.totalChunks, 0))
            }

            is FlyMessage.Chunk -> {
                pendingChunks[msg.id]?.add(body)
                val info = pendingFileInfos[msg.id] ?: return
                val received = pendingChunks[msg.id]?.size ?: 0
                updateFileProgress(msg.id, received.toFloat() / info.totalChunks, received)
            }

            is FlyMessage.FileDone -> {
                val chunks = pendingChunks.remove(msg.id) ?: return
                val info = pendingFileInfos.remove(msg.id) ?: return
                val fullData = chunks.fold(ByteArray(0)) { acc, chunk -> acc + chunk }
                val file = saveFile(info.name, fullData)
                updateTransfer(msg.id) { item ->
                    if (item is TransferItem.File) item.copy(localPath = file.absolutePath, progress = 1f)
                    else item
                }
                conn.send(FlyMessage.Ack(msg.id))
            }

            is FlyMessage.Logcat -> {
                val existing = _transfers.value.find { it.id == msg.id } as? TransferItem.LogcatStream
                if (existing != null) {
                    updateTransfer(msg.id) { item ->
                        if (item is TransferItem.LogcatStream) item.copy(lines = item.lines + msg.line)
                        else item
                    }
                } else {
                    prepend(TransferItem.LogcatStream(msg.id, deviceName, true, listOf(msg.line)))
                }
            }

            else -> {}
        }
    }

    suspend fun sendText(conn: PeerConnection, text: String) {
        val id = UUID.randomUUID().toString()
        conn.send(FlyMessage.Text(id, text))
        prepend(TransferItem.Text(id, deviceName(), false, text))
    }

    suspend fun sendUri(conn: PeerConnection, uri: Uri) {
        val id = UUID.randomUUID().toString()
        val name = resolveFileName(uri)
        val mime = context.contentResolver.getType(uri) ?: "application/octet-stream"
        val size = resolveFileSize(uri)
        val threshold = chunkThresholdBytes()

        if (size <= threshold) {
            val data = context.contentResolver.openInputStream(uri)?.readBytes() ?: return
            conn.send(FlyMessage.FileOffer(id, name, mime, size), data)
            prepend(TransferItem.File(id, deviceName(), false, name, mime, size, null, 1f))
        } else {
            val chunkSize = chunkSizeBytes()
            val totalChunks = ceil(size.toDouble() / chunkSize).toInt()
            conn.send(FlyMessage.FileStart(id, name, mime, size, totalChunks, chunkSize))
            prepend(TransferItem.File(id, deviceName(), false, name, mime, size, null, 0f, totalChunks, 0))
            withContext(Dispatchers.IO) {
                context.contentResolver.openInputStream(uri)?.use { stream ->
                    val buffer = ByteArray(chunkSize)
                    var index = 0
                    var bytesRead: Int
                    while (stream.read(buffer).also { bytesRead = it } != -1) {
                        conn.send(FlyMessage.Chunk(id, index), buffer.copyOf(bytesRead))
                        val received = index + 1
                        updateFileProgress(id, received.toFloat() / totalChunks, received)
                        index++
                    }
                }
            }
            conn.send(FlyMessage.FileDone(id))
            updateFileProgress(id, 1f, totalChunks)
        }
    }

    private fun saveFile(name: String, data: ByteArray): File {
        val dir = File(context.filesDir, "received").also { it.mkdirs() }
        return File(dir, name).also { it.writeBytes(data) }
    }

    private fun prepend(item: TransferItem) {
        _transfers.update { listOf(item) + it }
    }

    private fun updateFileProgress(id: String, progress: Float, received: Int = 0) {
        updateTransfer(id) { item ->
            if (item is TransferItem.File) item.copy(progress = progress, receivedChunks = received)
            else item
        }
    }

    private fun updateTransfer(id: String, transform: (TransferItem) -> TransferItem) {
        _transfers.update { list -> list.map { if (it.id == id) transform(it) else it } }
    }

    private fun resolveFileName(uri: Uri): String {
        if (uri.scheme == "content") {
            context.contentResolver.query(uri, arrayOf(OpenableColumns.DISPLAY_NAME), null, null, null)?.use { cursor ->
                if (cursor.moveToFirst()) {
                    val idx = cursor.getColumnIndex(OpenableColumns.DISPLAY_NAME)
                    if (idx >= 0) return cursor.getString(idx)
                }
            }
        }
        return uri.lastPathSegment ?: "file_${System.currentTimeMillis()}"
    }

    private fun resolveFileSize(uri: Uri): Long {
        if (uri.scheme == "content") {
            context.contentResolver.query(uri, arrayOf(OpenableColumns.SIZE), null, null, null)?.use { cursor ->
                if (cursor.moveToFirst()) {
                    val idx = cursor.getColumnIndex(OpenableColumns.SIZE)
                    if (idx >= 0 && !cursor.isNull(idx)) return cursor.getLong(idx)
                }
            }
        }
        return 0L
    }
}
